File: src/org/simbrain/network/gui/WandEventHandler.java
/*
 * Part of Simbrain--a java-based neural network kit
 * Copyright (C) 2005,2007 The Authors.  See http://www.simbrain.net/credits
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
package org.simbrain.network.gui;

import java.awt.event.InputEvent;
import java.awt.geom.Ellipse2D;
import java.awt.geom.Point2D;
import java.util.Collection;

import org.simbrain.network.gui.nodes.NeuronNode;
import org.simbrain.network.interfaces.Neuron;

import edu.umd.cs.piccolo.PCamera;
import edu.umd.cs.piccolo.PCanvas;
import edu.umd.cs.piccolo.PLayer;
import edu.umd.cs.piccolo.PNode;
import edu.umd.cs.piccolo.event.PDragSequenceEventHandler;
import edu.umd.cs.piccolo.event.PInputEvent;
import edu.umd.cs.piccolo.event.PInputEventFilter;
import edu.umd.cs.piccolo.util.PNodeFilter;

/**
 * Wand event handler.  Change activation when dragging over neurons.
 */
final class WandEventHandler extends PDragSequenceEventHandler {

    /** Bounds filter. */
    private final BoundsFilter boundsFilter;

    /** Network Panel. */
    private final NetworkPanel networkPanel;

    /**
     * Create a new selection event handler.
     */
    public WandEventHandler(NetworkPanel networkPanel) {
        super();
        boundsFilter = new BoundsFilter();
        setEventFilter(new ZoomEventFilter());
        this.networkPanel = networkPanel;
    }

    /** @see PDragSequenceEventHandler */
    public void mousePressed(final PInputEvent event) {
        super.mousePressed(event);
        networkPanel.setLastClickedPosition(event.getPosition());
        if (event.getPath().getPickedNode() instanceof PCamera) {
            networkPanel.setBeginPosition(event.getPosition());
        }
    }

    /** @see PDragSequenceEventHandler */
    public void mouseClicked(final PInputEvent event) {

        super.mouseClicked(event);

        if (event.getClickCount() != 1) {
            return;
        }

        PNode node = event.getPath().getPickedNode();
        if (node instanceof NeuronNode) {
            modifyNode((NeuronNode) node);
        }
    }

    /** @see PDragSequenceEventHandler */
    protected void startDrag(final PInputEvent event) {
        super.startDrag(event);
    }

    /** @see PDragSequenceEventHandler */
    protected void drag(final PInputEvent event) {

        super.drag(event);

        int radius = networkPanel.getEditMode().getWandRadius();

        // Create elliptical bounds
        Point2D position = event.getPosition();
        Ellipse2D.Double ellipse = new Ellipse2D.Double(position.getX()
                - radius / 2, position.getY() - radius / 2, radius, radius);
        boundsFilter.setEllipse(ellipse);

        Collection highlightedNodes = networkPanel.getCanvas().getLayer().getRoot()
                .getAllNodes(boundsFilter, null);

        // Auto-highlighter mode
        for (Object node : highlightedNodes) {
            if (node instanceof NeuronNode) {
                modifyNode((NeuronNode)node);
            }

        }

    }

    /** @see PDragSequenceEventHandler */
    protected void endDrag(final PInputEvent event) {
        super.endDrag(event);
    }

    /**
     * The wand "action" goes here.
     *
     * @param node node to act on
     */
    private void modifyNode(NeuronNode node) {
        Neuron neuron = node.getNeuron();
        neuron.setActivation(neuron.getUpperBound());
        node.update();
    }

    /**
     * Bounds filter.
     */
    private class BoundsFilter implements PNodeFilter {

        /** Bounds. */
        private Ellipse2D.Double ellipse;

        /**
         * Set the bounds for this bounds filter to <code>bounds</code>.
         *
         * @param ellipse
         *            bounds for this bounds filter
         */
        public void setEllipse(Ellipse2D.Double ellipse) {
            this.ellipse = ellipse;
        }

        /** @see PNodeFilter */
        public boolean accept(final PNode node) {
            boolean isPickable = node.getPickable();
            boolean boundsIntersects = ellipse.intersects(node
                    .getGlobalBounds());
            boolean isLayer = (node instanceof PLayer);
            boolean isCamera = (node instanceof PCamera);

            return (isPickable && boundsIntersects && !isLayer && !isCamera);
        }

        /** @see PNodeFilter */
        public boolean acceptChildrenOf(final PNode node) {
            boolean areChildrenPickable = node.getChildrenPickable();
            boolean isCamera = (node instanceof PCamera);
            boolean isLayer = (node instanceof PLayer);
            return (areChildrenPickable || isCamera || isLayer);
        }
    }

    /**
     * Selection event filter, accepts various mouse events, but only when
     * the network panel's edit mode is <code>EditMode.WAND</code>.
     */
    private class ZoomEventFilter
        extends PInputEventFilter {

        /**
         * Create a new selection event filter.
         */
        public ZoomEventFilter() {
            super(InputEvent.BUTTON1_MASK);
        }


        /** @see PInputEventFilter */
        public boolean acceptsEvent(final PInputEvent event, final int type) {

            EditMode editMode = networkPanel.getEditMode();

            if (editMode.isWand() && super.acceptsEvent(event, type)) {
                networkPanel.getTextHandle().stopEditing();
                return true;
            } else {
                return false;
            }
        }
    }

}

File: src/main/java/org/simbrain/network/gui/WandEventHandler.kt
package org.simbrain.network.gui

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.joinAll
import kotlinx.coroutines.launch
import kotlinx.coroutines.swing.Swing
import org.piccolo2d.PCamera
import org.piccolo2d.PLayer
import org.piccolo2d.PNode
import org.piccolo2d.event.PDragSequenceEventHandler
import org.piccolo2d.event.PInputEvent
import org.piccolo2d.event.PInputEventFilter
import org.piccolo2d.util.PNodeFilter
import org.simbrain.network.core.NetworkModel
import org.simbrain.network.gui.MouseEventHandler.MouseCursor
import org.simbrain.network.gui.dialogs.NetworkPreferences.wandPalette
import org.simbrain.network.gui.nodes.NeuronNode
import org.simbrain.network.gui.nodes.SynapseNode
import java.awt.event.InputEvent
import java.awt.geom.Ellipse2D

/**
 * Wand event handler. Applies the selected wand action when dragging over neurons.
 */
class WandEventHandler(val networkPanel: NetworkPanel) : PDragSequenceEventHandler() {
    /**
     * Bounds filter.
     */
    private val boundsFilter: BoundsFilter = BoundsFilter()

    /**
     * Create a new selection event handler.
     *
     * @param networkPanel
     */
    init {
        eventFilter = WandEventFilter()
    }

    /**
     * Tracks state for undo. Different actions may store different types.
     */
    private var undoState = mutableMapOf<Any, Any?>()

    /**
     * Tracks models that have already been operated on in this drag session.
     * This prevents repeat actions (e.g., randomizing the same neuron multiple times).
     */
    private val touchedModels = mutableSetOf<NetworkModel>()

    /**
     * Tracks pixels (neuron-array / weight-matrix cells) already operated on this drag session,
     * so a single drag doesn't randomize / increment the same pixel repeatedly.
     */
    private val touchedPixels = mutableSetOf<PixelTarget>()

    /**
     * Tracks pending apply jobs so we can await them before finalizing the action.
     */
    private val pendingJobs = mutableListOf<Job>()

    override fun mousePressed(event: PInputEvent) {
        super.mousePressed(event)
        undoState = mutableMapOf()
        touchedModels.clear()
        touchedPixels.clear()
        pendingJobs.clear()

        // Notify action that we're starting
        wandPalette.selectedAction?.beginAction(networkPanel)

        val node = event.path.pickedNode
        when (node) {
            is NeuronNode -> modifyModel(node.neuron)
            is SynapseNode -> modifyModel(node.synapse)
        }
    }

    override fun mouseReleased(event: PInputEvent?) {
        super.mouseReleased(event)
        val action = wandPalette.selectedAction
        val jobs = pendingJobs.toList()

        // Launch a coroutine to await all pending apply jobs, then finalize
        networkPanel.launch(Dispatchers.Swing) {
            jobs.joinAll()
            action?.endAction(networkPanel, undoState)
        }
    }

    override fun startDrag(event: PInputEvent?) {
        super.startDrag(event)
    }

    override fun drag(event: PInputEvent) {
        super.drag(event)

        val baseRadius = wandPalette.selectedAction?.radius ?: 40
        // Scale radius by inverse of zoom so cursor size matches effect area
        val viewScale = networkPanel.canvas.camera.viewScale
        val radius = baseRadius / viewScale

        // Create elliptical bounds
        val position = event.position
        val ellipse = Ellipse2D.Double(
            position.x - radius / 2,
            position.y - radius / 2,
            radius,
            radius
        )
        boundsFilter.setEllipse(ellipse)

        val highlightedNodes = networkPanel.canvas.layer.root.getAllNodes(boundsFilter, null)

        // Apply action to all neurons and synapses in bounds
        for (node in highlightedNodes) {
            when (node) {
                is NeuronNode -> modifyModel(node.neuron)
                is SynapseNode -> modifyModel(node.synapse)
            }
        }

        // Apply action to neuron-array / weight-matrix pixels under the cursor
        networkPanel.pixelsInGlobalEllipse(ellipse).forEach { pixel -> modifyPixel(pixel) }
    }

    override fun endDrag(event: PInputEvent?) {
        super.endDrag(event)
    }

    /**
     * Apply the selected wand action to a model.
     * Skips models that have already been operated on in this drag session.
     */
    private fun modifyModel(model: NetworkModel) {
        if (model in touchedModels) return
        touchedModels.add(model)
        val job = networkPanel.launch(Dispatchers.Swing) {
            wandPalette.selectedAction?.apply(model, networkPanel, undoState)
        }
        pendingJobs.add(job)
    }

    /**
     * Apply the selected wand action to a single array / matrix pixel.
     * Skips pixels already operated on in this drag session.
     */
    private fun modifyPixel(pixel: PixelTarget) {
        if (pixel in touchedPixels) return
        touchedPixels.add(pixel)
        val job = networkPanel.launch(Dispatchers.Swing) {
            wandPalette.selectedAction?.applyToPixel(pixel, networkPanel, undoState)
        }
        pendingJobs.add(job)
    }

    /**
     * Bounds filter.
     */
    private inner class BoundsFilter : PNodeFilter {
        /**
         * Bounds.
         */
        private var ellipse: Ellipse2D.Double? = null

        /**
         * Set the bounds for this bounds filter to `bounds`.
         *
         * @param ellipse bounds for this bounds filter
         */
        fun setEllipse(ellipse: Ellipse2D.Double) {
            this.ellipse = ellipse
        }

        /**
         * @param node
         * @return
         * @see PNodeFilter
         */
        override fun accept(node: PNode): Boolean {
            val isPickable = node.pickable
            val boundsIntersects = ellipse!!.intersects(node.globalBounds)
            val isLayer = (node is PLayer)
            val isCamera = (node is PCamera)

            return (isPickable && boundsIntersects && !isLayer && !isCamera)
        }

        override fun acceptChildrenOf(node: PNode): Boolean {
            val areChildrenPickable = node.childrenPickable
            val isCamera = (node is PCamera)
            val isLayer = (node is PLayer)
            return (areChildrenPickable || isCamera || isLayer)
        }
    }

    /**
     * Selection event filter, accepts various mouse events, but only when the
     * network panel's cursor is `MouseCursor.Wand`.
     */
    private inner class WandEventFilter : PInputEventFilter(InputEvent.BUTTON1_MASK) {
        override fun acceptsEvent(event: PInputEvent?, type: Int): Boolean {
            val mouseCursor = networkPanel.mouseCursor

            return mouseCursor === MouseCursor.Wand && super.acceptsEvent(event, type)
        }
    }
}
